signer/src/storage.rs

```diff
@@ -12,48 +12,57 @@ pub mod postgres;
 
 use std::future::Future;
 
+use blockstack_lib::chainstate::nakamoto::NakamotoBlock;
+use blockstack_lib::types::chainstate::StacksBlockId;
+
 /// Represents the ability to read data from the signer storage.
 pub trait DbRead {
     /// Read error.
     type Error: std::error::Error;
 
     /// Get the bitcoin block with the given block hash.
     fn get_bitcoin_block(
-        self,
+        &self,
```

Comment:
I'm alright with setting the receiver here, however to be cautious I think we should have it be mutable which I think would be necessary if we want to support transactions later down the line.

Reply:
You're right. If we want to implement `DbRead` or `DbWrite` for `sqlx::Transaction` (or something wrapping it) then we'll need to change these traits to require `&mut self`. If we want transaction support I'm okay with changing the API. But we're okay if we just want transaction functionality within the function implementation since we only need a reference to a `PgPool` to create a `sqlx::Transaction`.

---

My personal preferences (loosely held) is to support a (wrapped) `sqlx::Transaction` implementation but only for testing purposes. But this is to avoid `sqlx` macros like `#[sqlx::test]` and `sqlx::query!`. Basically, I'm still conflicted about the utility of the statically checked queries since it doesn't help more than an integration test would (which I think we should always add). Hopefully this helps explain why I didn't use the `sqlx::query` in this PR.

Reply:
Makes sense. I'm not sure if we need transactional support, but I'd err on the side of caution and use `&mut` here so we retain the ability to use it and not starting to rely on these references to be immutable.

Reply:
> Basically, I'm still conflicted about the utility of the statically checked queries since it doesn't help more than an integration test would (which I think we should always add). Hopefully this helps explain why I didn't use the sqlx::query in this PR.

That's a fair point. What I like about the queries is that I can trust the query to be correct without having to look if there is an integration test, which makes it much faster to verify the code when reading it. I agree that having an integration test provides strictly more utility, and the macros shouldn't substitute it. Though you will need the database up when running the integration test so at that point it should be fairly easy to also run `cargo sqlx prepare`, so it shouldn't be too much extra effort to use the macros either.

Reply:
> Though you will need the database up when running the integration test so at that point it should be fairly easy to also run `cargo sqlx prepare`, so it shouldn't be too much extra effort to use the macros either.

True, it's not much extra effort.

---

> What I like about the queries is that I can trust the query to be correct without having to look if there is an integration test, which makes it much faster to verify the code when reading it. 

Yeah for some queries this makes sense and works perfectly, but sometimes not so much. For the queries added here, it's not possible to know if things will work as intended without an integration test.

For example, in this PR one of my original queries looked like this:
```sql
INSERT INTO sbtc_signer.stacks_transactions (txid, block_hash)
SELECT
    txid
  , block_id
FROM JSONB_TO_RECORDSET($1::JSONB) AS x(
    txid        BYTEA
  , block_id    BYTEA
)
ON CONFLICT DO NOTHING
```

Turns out that this will pass any type checks and run successfully but won't do what we want it to do (probably because we go through the `JSONB_TO_RECORDSET` function). Because things don't always work as expected, I'm wary of relying solely on type checks, even for simple queries.

Reply:
True. Type checks can tell you that something is wrong but it won't guarantee that it is correct.

Reply:
Yeah, I know you know that. I just wanted to highlight a specific simple example where things don't quite work.

Reply:
Yeah it's a good example, thanks for highlighting!